### ISSUE TITLE
Request: Migration documentation update

### ISSUE THREAD

**Issue by dionsnoeijen** [web-token/jwt-framework#70]
It would be great if the documentation was a little more clear on how to make an implementation.

Right now I'm using this library as an example for an AWS Cognito implementation. 
https://github.com/pmill/aws-cognito/blob/master/src/CognitoClient.php
It's using the old spomky lib, so I thought I would use the new one. But I find it very hard to follow the documentation.

For example, some code from the pmill cognito library:

```
    /**
     * @param CacheItemPoolInterface $cache
     */
    public function downloadJwtWebKeys(CacheItemPoolInterface $cache = null)
    {
        $url = sprintf(
            'https://cognito-idp.%s.amazonaws.com/%s/.well-known/jwks.json',
            $this->region,
            $this->userPoolId
        );
        $this->jwtWebKeys = JWKFactory::createFromJKU($url, false, $cache);
    }
```

What replaces the create from JKU? I Suppose it's this configuration: https://web-token.spomky-labs.com/bundle/jwk/keysets.html

```
jose:
    keys:
        key_name:
            x5u: # Method
                url: 'https://www.googleapis.com/oauth2/v1/certs'
```

If so, how is it used when I want to have a method like this one: (also from the pmill cognito library)

```
public function verifyAccessToken($accessToken)
    {
        $signatureIndex = null;
        $loader = new Loader();
        $jwt = $loader->loadAndVerifySignatureUsingKeySet($accessToken, $this->jwtWebKeys, ['RS256'], $signatureIndex);
        /** @var array $jwtPayload */
        $jwtPayload = $jwt->getPayload();
        $expectedIss = sprintf('https://cognito-idp.%s.amazonaws.com/%s', $this->region, $this->userPoolId);
        if ($jwtPayload['iss'] !== $expectedIss) {
            throw new TokenVerificationException('invalid iss');
        }
        if ($jwtPayload['token_use'] !== 'access') {
            throw new TokenVerificationException('invalid token_use');
        }
        if ($jwtPayload['exp'] < time()) {
            throw new TokenExpiryException('invalid exp');
        }
        return $jwtPayload['username'];
    }
```
Above I see a loader, that is verifying the signature using a key set. Is it done somewhere along the lines of this?

    public function createJwtWebKeys(): JWKSet
    {
        return JWKSet::createFromJson('{"keys": /* removed */');
    }

    public function verifyAccessToken(string $accessToken): bool
    {
        // Signature?
        // The verifier is created like this: $this->verifier = $jwsVerifierFactory->create(['HS256']);
        // $this->jwkSet is set with createJwtWebKeys by manually pasting the response from aws from: https://cognito-idp.%s.amazonaws.com/%s
        return $this->verifier->verifyWithKeySet(JWS::create($accessToken), $this->jwkSet, 0);
    }

Any info on migrating a situation like this would be great!

Thanks!

**Comment by Spomky** [comment]
Hi,

You are right, the use of the bundle is not clear as it should be.
The easiest way in your case is to inject the JKUFactory service into your `CognitoClient`.
Be sure you have the following dependencies installed:

* (`web-token/jwt-bundle` and `web-token/jwt-key-mgmt`) or `web-token/jwt-framework`
* `php-http/httplug-bundle` and [at least one adapter](http://docs.php-http.org/en/latest/clients.html#clients-adapters) (I will use `php-http/guzzle6-adapter` here)

Do not forget to enable the associated sf bundles:
```php
new Http\HttplugBundle\HttplugBundle(),
new Jose\Bundle\JoseFramework\JoseFrameworkBundle(),
```

The configuration should be as follow:

```yaml
jose:
    jku_factory:
        enabled: true # We enable the JKU factory
        client: 'httplug.client.my_client' # We indicate the Httplug client to use
        request_factory: 'AppBundle\Service\RequestFactory' # See hereafter the corresponding class

httplug:
    plugins:
        cache: # We use the cache plugin
            cache_pool: 'cache.app' # We use the PSR-6 Cache service of the application
            config:
                default_ttl: 1800 # TTL set to 30 min
    clients:
        my_client: # Our client based on Guzzle 6. The corresponding service will be `httplug.client.my_client`
            factory: 'httplug.factory.guzzle6'
            plugins: ['httplug.plugin.cache'] # We enable the cache plugin for that client
```

The `RequestFactory` class is just a simple clas that create Psr7 Requests on demand.

```php
<?php
# app/AppBundle/Service/RequestFactory.php

namespace AppBundle\Service;

use GuzzleHttp\Psr7\Request;
use Http\Message\RequestFactory as Psr7RequestFactory;

final class RequestFactory implements Psr7RequestFactory
{
    /**
     * {@inheritdoc}
     */
    public function createRequest($method, $uri, array $headers = [], $body = null, $protocolVersion = '1.1')
    {
        return new Request($method, $uri, $headers, $body, $protocolVersion);
    }
}
```

When done, the `JKUFactory` service will be available and will retrieve the key set using the URL (or the cache if available).

```php
use Jose\Component\KeyManagement\JKUFactory;
...
class CognitoClient
{
    ...
    private $jkuFactory;

    public function __construct(CognitoIdentityProviderClient $client, JKUFactory $jkuFactory)
    {
        $this->client = $client;
    }
    ...
    public function downloadJwtWebKeys()
    {
        $url = sprintf(
            'https://cognito-idp.%s.amazonaws.com/%s/.well-known/jwks.json',
            $this->region,
            $this->userPoolId
        );
        $this->jwtWebKeys = $this->jkuFactory->loadFromUrl($url);
    }
}

Let me know if it solves your issue. In any case I will update the documentation to explain all the ways you have to load and use key sets (another way is the use of the [jose app](https://github.com/web-token/jwt-app) and env vars).

**Comment by Spomky** [comment]
Hi @dionsnoeijen,

I have just updated [the documentation](https://web-token.spomky-labs.com/migration/spomky-labs-jose/jwkset.html) related to that topic.

**Comment by dionsnoeijen** [comment]
Thank you @Spomky, I will be working on this again in a few hours. When necessary I will keep you posted.

**Comment by Spomky** [comment]
Closing as done.
Feel free to reopen if needed.
